File: [K]graph/[K]graph-kahn/207-course-schedule.java
class Solution {
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        HashMap<Integer, List<Integer>> graph = new HashMap<>();
        HashMap<Integer, Integer> indegrees = new HashMap<>();
        for (int i = 0; i < numCourses; i++) {
            graph.put(i, new ArrayList<>());
            indegrees.put(i, 0);
        }
        for (int[] prereq: prerequisites) {
            int q = prereq[0];
            int p = prereq[1];
            graph.get(p).add(q);
            indegrees.put(q, indegrees.get(q) + 1);
        }

        ArrayDeque<Integer> queue = new ArrayDeque<>();
        for (Integer node: indegrees.keySet()) {
            if (Objects.equals(indegrees.get(node), 0)) {
                queue.offer(node);
            }
        }
        ArrayList<Integer> res = new ArrayList<>();
        while (!queue.isEmpty()) {
            Integer node = queue.poll();
            res.add(node);
            for (Integer neighbor: graph.get(node)) {
                indegrees.put(neighbor, indegrees.get(neighbor) - 1);
                if (Objects.equals(indegrees.get(neighbor), 0)) {
                    queue.offer(neighbor);
                }
            }
        }
        return res.size() == numCourses;
    }
}

// time O(V+E)
// space O(V+E), due to building graph
// using graph and kahn and topological sort

File: [K]graph/[K]graph-kahn/207-course-schedule.py
from collections import deque
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        graph = [[] for _ in range(numCourses)]
        indegrees = [0 for _ in range(numCourses)]
        for q, p in prerequisites:
            graph[p].append(q)
            indegrees[q] += 1

        queue = deque([])
        for i, indegree in enumerate(indegrees):
            if indegree == 0:
                queue.append(i)

        res = []
        while queue:
            node = queue.popleft()
            res.append(node)
            for neighbor in graph[node]:
                indegrees[neighbor] -= 1
                if indegrees[neighbor] == 0:
                    queue.append(neighbor)

        return len(res) == numCourses
    
# time O(V+E)
# space O(V+E), due to building graph
# using graph and kahn and topological sort

class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        graph = [[] for _ in range(numCourses)]
        for q, p in prerequisites:
            graph[p].append(q)
        
        node_flag = [0 for _ in range(numCourses)]
        cyclic = False

        def dfs(node):
            nonlocal cyclic
            node_flag[node] = 1
            for neighbor in graph[node]:
                if node_flag[neighbor] == 0:
                    dfs(neighbor)
                elif node_flag[neighbor] == 1:
                    cyclic = True
            node_flag[node] = 2

        for node in range(numCourses):
            if node_flag[node] == 0:
                dfs(node)
                
        return not cyclic

# time O(V+E)
# space O(V+E)
# using graph and dfs and detecting cycles
'''
# 0 not visited
# 1 visiting
# 2 completed visited
'''